stan/math/prim/prob/hmm_marginal_lpdf.hpp

```diff
@@ -0,0 +1,187 @@
+ * @tparam T_Gamma type of the transition matrix
+ * @tparam T_rho type of the initial guess vector
+ *
+ * @param[in] log_omega log matrix of observational densities.
+ *              The (i, j)th entry corresponds to the
+ *              density of the ith observation, y_i,
+ *              given x_i = j.
+ * @param[in] Gamma transition density between hidden states.
+ *              The (i, j)th entry is the probability that x_n = j,
+ *              given x_{n - 1} = i. The rows of Gamma are simplexes.
+ * @param[in] rho initial state
+ * @throw `std::domain_error` if Gamma is not square.
+ * @throw `std::invalid_argument` if the size of rho is not
+ * the number of rows of Gamma.
+ * @throw `std::domain_error` if rho is not a simplex.
+ * @throw `std::domain_error` if the rows of Gamma are
+ * not a simplex.
+ * @return log marginal density.
+ */
+template <typename T_omega, typename T_Gamma, typename T_rho>
+inline auto hmm_marginal_lpdf(
+    const Eigen::Matrix<T_omega, Eigen::Dynamic, Eigen::Dynamic>& log_omegas,
+    const Eigen::Matrix<T_Gamma, Eigen::Dynamic, Eigen::Dynamic>& Gamma,
+    const Eigen::Matrix<T_rho, Eigen::Dynamic, 1>& rho) {
+  using T_partial_type = partials_return_t<T_omega, T_Gamma, T_rho>;
+  using eig_matrix_partial
+      = Eigen::Matrix<T_partial_type, Eigen::Dynamic, Eigen::Dynamic>;
+  using eig_vector_partial = Eigen::Matrix<T_partial_type, Eigen::Dynamic, 1>;
+  int n_states = log_omegas.rows();
+  int n_transitions = log_omegas.cols() - 1;
+
+  check_square("hmm_marginal_lpdf", "Gamma", Gamma);
+  check_consistent_size("hmm_marginal_lpdf", "Gamma", row(Gamma, 1), n_states);
+  check_consistent_size("hmm_marginal_lpdf", "rho", rho, n_states);
+  check_simplex("hmm_marginal_lpdf", "rho", rho);
+  for (int i = 0; i < Gamma.rows(); ++i) {
+    check_simplex("hmm_marginal_lpdf", "Gamma[i, ]", row(Gamma, i + 1));
+  }
+
+  operands_and_partials<Eigen::Matrix<T_omega, Eigen::Dynamic, Eigen::Dynamic>,
+                        Eigen::Matrix<T_Gamma, Eigen::Dynamic, Eigen::Dynamic>,
+                        Eigen::Matrix<T_rho, Eigen::Dynamic, 1> >
+      ops_partials(log_omegas, Gamma, rho);
+
+  eig_matrix_partial alphas(n_states, n_transitions + 1);
+  eig_vector_partial alpha_log_norms(n_transitions + 1);
+  eig_matrix_partial omegas;
+  auto Gamma_val = value_of(Gamma).eval();
+
+  // compute the density using the forward algorithm.
+  {
+    const auto log_omegas_val = value_of(log_omegas).eval();
+    const auto rho_val = value_of(rho).eval();
+    omegas = log_omegas_val.array().exp();
+    const int n_states = log_omegas_val.rows();
+    const int n_transitions = log_omegas_val.cols() - 1;
+
+    alphas.col(0) = omegas.col(0).cwiseProduct(rho_val);
+
+    const auto norm = alphas.col(0).maxCoeff();
+    alphas.col(0) /= norm;
+    alpha_log_norms(0) = log(norm);
+
+    for (int n = 0; n < n_transitions; ++n) {
+      alphas.col(n + 1) = omegas.col(n + 1).cwiseProduct(Gamma_val.transpose()
```

Comment:
> Would it be more efficient to .eval() the transpose outside the loop? If the componentwise product here is by column, then that'd make all the cwiseProduct better in terms of memory locality.

Definitely

Reply:
Does the transposition come at any cost? We don't create a new object do we, if we're using lazy operations?

Reply:
We do create a new object for transpose and have to pay the cost of the constructor.  But that object doesn't copy the data, it only holds a reference to it.  So these costs are low.

Consider a matrix multiply `A * B`.  If you store `A` in column major mode, then every time you access a row (once for each column of `B`), you have to do a non-memory-local traversal.  It can be faster to use an efficient transpose, do it once, then have things be memory local afterward.  I believe that Eigen does this internally for multiplies if the matrices are big enough.